src/dbnode/storage/index.go

```diff
@@ -944,43 +944,73 @@ func (i *nsIndex) flushableBlocks(
 	if !i.isOpenWithRLock() {
 		return nil, errDbIndexUnableToFlushClosed
 	}
+	// NB(bodu): We read index info files once here to avoid re-reading all of them
+	// for each block.
+	fsOpts := i.opts.CommitLogOptions().FilesystemOptions()
+	infoFiles := i.readIndexInfoFilesFn(
+		fsOpts.FilePathPrefix(),
+		i.nsMetadata.ID(),
+		fsOpts.InfoReaderBufferSize(),
+	)
 	flushable := make([]index.Block, 0, len(i.state.blocksByTime))
-	for _, block := range i.state.blocksByTime {
-		canFlush, err := i.canFlushBlock(block, shards, flushType)
+
+	now := i.nowFn()
+	earliestBlockStartToRetain := retention.FlushTimeStartForRetentionPeriod(i.retentionPeriod, i.blockSize, now)
+	currentBlockStart := now.Truncate(i.blockSize)
+	// Check for flushable blocks by iterating through all block starts w/in retention.
+	for blockStart := earliestBlockStartToRetain; blockStart.Before(currentBlockStart); blockStart = blockStart.Add(i.blockSize) {
+		canFlush, err := i.canFlushBlockWithRLock(infoFiles, now, blockStart, shards, flushType)
 		if err != nil {
 			return nil, err
 		}
 		if !canFlush {
 			continue
 		}
+		// Ensure all flushable blocks exist.
+		block, err := i.ensureBlockPresent(blockStart)
+		if err != nil {
+			return nil, err
+		}
 		flushable = append(flushable, block)
 	}
 	return flushable, nil
 }
 
-func (i *nsIndex) canFlushBlock(
-	block index.Block,
+func (i *nsIndex) canFlushBlockWithRLock(
+	infoFiles []fs.ReadIndexInfoFileResult,
+	startTime time.Time,
+	blockStart time.Time,
 	shards []databaseShard,
 	flushType series.WriteType,
 ) (bool, error) {
-	// Check the block needs flushing because it is sealed and has
-	// any mutable/cold mutable segments that need to be evicted from memory.
 	switch flushType {
 	case series.WarmWrite:
-		if !block.IsSealed() || !block.NeedsMutableSegmentsEvicted() {
+		// NB(bodu): We should always attempt to warm flush sealed blocks to disk if
+		// there doesn't already exist data on disk. We're checking this instead of
+		// `block.NeedsMutableSegmentsEvicted()` since bootstrap writes for cold block starts
+		// get marked as warm writes if there doesn't already exist data on disk and need to
+		// properly go through the warm flush lifecycle.
+		isSealed := !blockStart.After(i.lastSealableBlockStart(startTime))
+		if !isSealed || i.hasWarmFlushedToDisk(infoFiles, blockStart) {
```

Comment:
Hm, shouldn't this be "if not sealed || not warm flushed to disk { return false }"?  

Seems like `i.hasWarmFlushedToDisk(infoFiles, blockStart)` needs to be inverted?